tools/workflowresultparser/main.go

```diff
@@ -0,0 +1,205 @@
+package main
+
+import (
+	"encoding/json"
+	"flag"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"regexp"
+	"time"
+)
+
+type Job struct {
+	Name  string `json:"name"`
+	Steps []Step `json:"steps"`
+	URL   string `json:"html_url"`
+}
+
+type Step struct {
+	Name       string `json:"name"`
+	Conclusion string `json:"conclusion"`
+}
+
+type GitHubResponse struct {
+	TotalCount int   `json:"total_count"`
+	Jobs       []Job `json:"jobs"`
+}
+
+type ParsedResult struct {
+	Conclusion string `json:"conclusion"`
+	Cap        string `json:"cap"`
+	URL        string `json:"html_url"`
+}
+
+type ResultsMap map[string][]ParsedResult
+
+type HTTPClient interface {
+	Do(req *http.Request) (*http.Response, error)
+}
+
+func fetchGitHubJobs(apiURL, token string, client HTTPClient) ([]Job, error) {
+	var allJobs []Job
+
+	for {
+		req, err := http.NewRequest("GET", apiURL, nil)
+		if err != nil {
+			return nil, fmt.Errorf("error creating HTTP request: %w", err)
+		}
+		req.Header.Set("Authorization", "Bearer "+token)
+
+		resp, err := client.Do(req)
+		if err != nil {
+			return nil, fmt.Errorf("error making HTTP request: %w", err)
+		}
+
+		if resp.StatusCode != http.StatusOK {
+			return nil, fmt.Errorf("GitHub API request failed with status: %s", resp.Status)
+		}
+
+		body, err := io.ReadAll(resp.Body)
+		if err != nil {
+			return nil, fmt.Errorf("error reading response body: %w", err)
+		}
+
+		_ = resp.Body.Close()
+
+		var githubResponse GitHubResponse
+		err = json.Unmarshal(body, &githubResponse)
+		if err != nil {
+			return nil, fmt.Errorf("error unmarshalling JSON response: %w", err)
+		}
+
+		allJobs = append(allJobs, githubResponse.Jobs...)
+
+		re := regexp.MustCompile(`&page=(\d+)`)
+		matches := re.FindStringSubmatch(apiURL)
+		pageNum := 0
+		if len(matches) > 1 {
+			_, err = fmt.Sscanf(matches[1], "%d", &pageNum)
+			if err != nil {
+				return nil, fmt.Errorf("error parsing page number: %w", err)
+			}
+		}
+
+		if (len(githubResponse.Jobs) < 100) || (len(githubResponse.Jobs) == 100 && githubResponse.TotalCount/100 == pageNum) {
+			break
+		}
+
+		if pageNum == 0 {
+			apiURL = apiURL + "&page=2"
+		} else {
+			apiURL = re.ReplaceAllString(apiURL, fmt.Sprintf("&page=%d", pageNum+1))
+		}
+	}
+
+	return allJobs, nil
+}
+
+func parseResults(jobNameRegex, workflowRunID *string, jobs []Job) ([]ParsedResult, error) {
+	var parsedResults []ParsedResult
+	re := regexp.MustCompile(*jobNameRegex)
+	for _, job := range jobs {
+		if re.MatchString(job.Name) {
+			for _, step := range job.Steps {
+				if step.Name == "Run Tests" {
+					conclusion := ":x:"
+					if step.Conclusion == "success" {
+						conclusion = ":white_check_mark:"
+					}
+					captureGroup := re.FindStringSubmatch(job.Name)[1]
```

Comment:
Can we have an empty capture group here?

Reply:
I was also worried about it, but I have a test that doesn't match anything and nothing panics, which is a bit surprising, tbh

I was thinking also about adding a input job regex validation to make sure it contains at least 1 capture group